### PR TITLE
feat(smart-contracts): add stableswap support for provide liquidity

### PR DESCRIPTION
**Pull request by nseguias**
## Description and Motivation

This PR adds support for `StableSwap` pools when providing liquidity through the Pool Manager.

## Related Issues

#311 

---

## Checklist:

- [x] I have read [Migaloo's contribution guidelines](https://github.com/White-Whale-Defi-Platform/migaloo-core/blob/main/docs/CONTRIBUTING.md).
- [x] My pull request has a sound title and description (not something vague like `Update index.md`)
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation.
- [x] The code is formatted properly `cargo fmt --all --`.
- [x] Clippy doesn't report any issues `cargo clippy -- -D warnings`.
- [x] I have regenerated the schemas if needed `cargo schema`.

### CODE REVIEW

File: contracts/liquidity_hub/pool-manager/src/helpers.rs [hunk 1]
```diff
@@ -487,7 +490,7 @@ pub fn assert_slippage_tolerance(
         }
 
         let one_minus_slippage_tolerance = Decimal256::one() - slippage_tolerance;
-        let deposits: [Uint256; 2] = [deposits[0].into(), deposits[1].into()];
+        let deposits: [Uint256; 2] = [deposits[0].amount.into(), deposits[1].amount.into()];
```
Comment (kerber0x) [inline contracts/liquidity_hub/pool-manager/src/helpers.rs:493]: how is this handled if it's a 3pool?

File: contracts/liquidity_hub/pool-manager/src/helpers.rs [hunk 2]
```diff
@@ -662,3 +665,120 @@ pub fn get_asset_indexes_in_pool(
...
+                    .unwrap()
+                    .checked_div(amount.into())
+                    .unwrap();
+                d_prev = d;
+                d = compute_next_d(amp_factor, d, d_prod, sum_x, n_coins).unwrap();
+                // Equality with the precision of 1
+                if d > d_prev {
+                    if d.checked_sub(d_prev).unwrap() <= Uint256::one() {
+                        break;
+                    }
```
[the remark this answers is not in the document]
Comment (kerber0x) [reply contracts/liquidity_hub/pool-manager/src/helpers.rs:705]: +1

File: contracts/liquidity_hub/pool-manager/src/helpers.rs [hunk 3]
```diff
@@ -662,3 +665,120 @@ pub fn get_asset_indexes_in_pool(
...
+                }
+            }
+        }
+        Some(d)
+    }
+}
+
+// TODO: handle unwraps properly
+#[allow(clippy::unwrap_used)]
+fn compute_next_d(
```
Comment (kerber0x) [inline contracts/liquidity_hub/pool-manager/src/helpers.rs:717]: for all these functions copied from the 3pool `curve.rs`, copy pls the uni tests in it

File: contracts/liquidity_hub/pool-manager/src/liquidity/commands.rs [hunk 4]
```diff
@@ -275,10 +264,43 @@ pub fn provide_liquidity(
                     amount
                 }
             }
-            PoolType::StableSwap { amp: _ } => {
-                // TODO: Handle stableswap
+            PoolType::StableSwap { amp: amp_factor } => {
+                if total_share == Uint128::zero() {
+                    // Make sure at least MINIMUM_LIQUIDITY_AMOUNT is deposited to mitigate the risk of the first
+                    // depositor preventing small liquidity providers from joining the pool
+                    let min_lp_token_amount = MINIMUM_LIQUIDITY_AMOUNT * Uint128::from(3u8);
```
Comment (kerber0x) [inline contracts/liquidity_hub/pool-manager/src/liquidity/commands.rs:271]: What is the hardcoded `Uint128::from(3u8)`? would that be `N_COINS`?

File: contracts/liquidity_hub/pool-manager/src/liquidity/commands.rs [hunk 5]
```diff
@@ -275,10 +264,43 @@ pub fn provide_liquidity(
...
+                    let share = Uint128::try_from(compute_d(amp_factor, &deposits).unwrap())?
+                        .checked_sub(min_lp_token_amount)
+                        .map_err(|_| {
+                            ContractError::InvalidInitialLiquidityAmount(min_lp_token_amount)
+                        })?;
+
```
Comment (kerber0x) [inline contracts/liquidity_hub/pool-manager/src/liquidity/commands.rs:277]: No LP tokens are being minted to the contract, i.e. the `MINIMUM_LIQUIDITY_AMOUNT`. See the case for `PoolType::ConstantProduct` when the `total_share == Uint128::zero()`

File: contracts/liquidity_hub/pool-manager/src/tests/integration_tests.rs [hunk 6]
```diff
@@ -2023,6 +2023,8 @@ mod swapping {
         );
     }
 
+    // TODO: make sure stableswap test works when pool has ONLY 2 assets
+
```
Comment (kerber0x) [inline contracts/liquidity_hub/pool-manager/src/tests/integration_tests.rs:2027]: Have tests for the stableswap with 2 assets being done? There are still some hardcoded 3s around, makes be feel it won't work with 2.

File: contracts/liquidity_hub/pool-manager/src/tests/integration_tests.rs [hunk 7]
```diff
@@ -3487,6 +3491,500 @@ mod provide_liquidity {
...
+                                _ => {}
+                            }
+                        }
+                    }
+                }
+                // Because the Pool was created and 1_000_000 of each token has been provided as liquidity
+                // Assuming no fees we should expect a small swap of 1000 to result in not too much slippage
+                // Expect 1000 give or take 0.002 difference
+                // Once fees are added and being deducted properly only the "0.002" should be changed.
+                assert_approx_eq!(
```
Comment (kerber0x) [inline contracts/liquidity_hub/pool-manager/src/tests/integration_tests.rs:3631]: `assert_approx_eq ` i like this 👀

**Review by kerber0x — CHANGES_REQUESTED** [review]
It looks good overall, however i think it won't work for a stableswap with 2 assets as there's `N_COINS` and other hardcoded value equal to `3`. Also, the unit tests for computing the d and other values for the stableswap were not transferred from the stableswap code.

I'd suggest all these compute_d and related methods to be moved to a file called math.

File: contracts/liquidity_hub/pool-manager/src/helpers.rs [hunk 8]
```diff
@@ -18,15 +18,15 @@ const NEWTON_ITERATIONS: u64 = 32;
 
 // todo isn't this for the 3pool? shouldn't it be 3
 // the number of assets in the pool
-const N_COINS: Uint256 = Uint256::from_u128(2);
+pub const N_COINS: u8 = 3;
```
[the remark this answers is not in the document]
Comment (nseguias) [reply contracts/liquidity_hub/pool-manager/src/helpers.rs:21]: using `Uint256::from(pool_info.assets.len() as u128)` now

File: contracts/liquidity_hub/pool-manager/src/liquidity/commands.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (nseguias) [reply contracts/liquidity_hub/pool-manager/src/liquidity/commands.rs:271]: I actually change it to work with N amount of assets `let min_lp_token_amount = MINIMUM_LIQUIDITY_AMOUNT * Uint128::from(pool_assets.len() as u128)`

File: contracts/liquidity_hub/pool-manager/src/helpers.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (nseguias) [reply contracts/liquidity_hub/pool-manager/src/helpers.rs]: All 3pool tests && proptests have been refactored and integrated into the pool-manager

File: contracts/liquidity_hub/pool-manager/src/helpers.rs (hunk 3 above)
Comment (nseguias) [reply contracts/liquidity_hub/pool-manager/src/helpers.rs:731]: done

File: contracts/liquidity_hub/pool-manager/src/liquidity/commands.rs (hunk 4 above)
Comment (nseguias) [reply contracts/liquidity_hub/pool-manager/src/liquidity/commands.rs]: fixed

File: contracts/liquidity_hub/pool-manager/src/tests/integration_tests.rs (hunk 6 above)
Comment (nseguias) [reply contracts/liquidity_hub/pool-manager/src/tests/integration_tests.rs:2027]: it has been tested with `amp: 100`

File: contracts/liquidity_hub/pool-manager/src/tests/integration_tests.rs [hunk 9]
```diff
@@ -3487,6 +3491,500 @@ mod provide_liquidity {
...
+            creator.clone(),
+            "whale-uluna-uusd".to_string(),
+            None,
+            None,
+            None,
+            None,
+            vec![
+                Coin {
+                    denom: "uwhale".to_string(),
+                    amount: Uint128::from(MINIMUM_LIQUIDITY_AMOUNT),
```
[the remark this answers is not in the document]
Comment (nseguias) [reply contracts/liquidity_hub/pool-manager/src/tests/integration_tests.rs:3873]: we're doing a swap a lil bit further down.

File: contracts/liquidity_hub/pool-manager/src/liquidity/commands.rs (hunk 5 above)
Comment (kerber0x) [reply contracts/liquidity_hub/pool-manager/src/liquidity/commands.rs:270]: Fixed in latest commit

**Review by kerber0x — APPROVED** [review]

**Pull request merged**